File: src/tag_dictionary.rs
use godot::prelude::*;

pub const SPLIT_CHAR: char = '.';

#[derive(GodotClass)]
#[class(tool, init, base = Resource)]
pub struct TagDictionary {
    base: Base<Resource>,
    #[export]
    tags: PackedStringArray,
}

#[godot_api]
impl TagDictionary {
    #[func]
    pub fn add_tag(&mut self, tag: GString) {
        if !self.tags.contains(&tag) {
            self.tags.push(tag);
            self.base_mut().emit_changed();
        }
    }

    #[func]
    pub fn add_tags(&mut self, tags: PackedStringArray) {
        let mut count = 0;

        for tag in tags.as_slice() {
            if !self.tags.contains(&tag) {
                self.tags.push(tag.clone());
                count += 1;
            }
        }

        if count > 0 {
            self.base_mut().emit_changed();
        }
    }

    #[func]
    pub fn count(&self) -> u64 {
        self.tags.len() as u64
    }

    #[func]
    pub fn find(&self, predicate: Callable) -> PackedStringArray {
        let mut result = PackedStringArray::new();

        for tag in self.tags.as_slice() {
            let mut args = VariantArray::new();

            args.push(tag.to_variant());

            if predicate.callv(args).booleanize() {
                result.push(tag.clone());
            }
        }

        result
    }

    #[func]
    pub fn get_tags_from_path(&self, path: GString) -> PackedStringArray {
        let mut out = PackedStringArray::new();

        for tag in self.tags.as_slice() {
            if tag.to_string().starts_with(path.to_string().as_str()) {
                out.push(tag.clone());
            }
        }

        out
    }

    #[func]
    pub fn get_tree(&self) -> Dictionary {
        let root = Dictionary::new();
        let mut tags = self.tags.clone();

        tags.sort();

        for tag in tags.as_slice() {
            let cloned = tag.clone().to_string();
            let chunks = cloned.split(SPLIT_CHAR);
            let mut current = root.clone();

            for chunk in chunks {
                // if current has the key, pass, otherwise current is a new dictionary
                match current.get(chunk) {
                    Some(_dict) => {
                        current = _dict.to();
                    }
                    None => {
                        let new_dict = Dictionary::new();
                        current.set(chunk, new_dict.clone());
                        current = new_dict;
                    }
                }
            }
        }

        root
    }

    #[func]
    pub fn has_path(&self, path: String) -> bool {
        self.tags
            .as_slice()
            .iter()
            .any(|tag| tag.to_string().starts_with(&path))
    }

    #[func]
    pub fn has_tag(&self, tag: GString) -> bool {
        self.tags.contains(&tag)
    }

    #[func]
    pub fn has_some_tags(&self, tags: PackedStringArray) -> bool {
        tags.as_slice().iter().any(|tag| self.tags.contains(&tag))
    }

    #[func]
    pub fn has_none_tags(&self, tags: PackedStringArray) -> bool {
        tags.as_slice().iter().all(|tag| !self.tags.contains(&tag))
    }

    #[func]
    pub fn none(&self, predicate: Callable) -> bool {
        self.tags.as_slice().iter().all(|tag| {
            let mut args = VariantArray::new();

            args.push(tag.to_variant());

            !predicate.callv(args).booleanize()
        })
    }

    #[func]
    pub fn replace_tag(&mut self, old_tag: String, new_tag: String) {
        if let Some(index) = self
            .tags
            .as_slice()
            .iter()
            .position(|t| old_tag == t.to_string())
        {
            self.tags[index] = new_tag.into();
            self.base_mut().emit_changed();
        }
    }

    #[func]
    pub fn replace_tags(&mut self, old_tags: PackedStringArray, new_tags: PackedStringArray) {
        let mut count = 0;

        for old_tag in old_tags.as_slice() {
            if let Some(index) = self.tags.as_slice().iter().position(|t| old_tag == t) {
                if let Some(tag) = new_tags.get(index) {
                    self.tags[index] = tag;
                }
                count += 1;
            }
        }

        if count > 0 {
            self.base_mut().emit_changed();
        }
    }

    #[func]
    pub fn remove_tag(&mut self, tag: GString) {
        if let Some(index) = self.tags.as_slice().iter().position(|t| tag.eq(t)) {
            self.tags.remove(index);
            self.base_mut().emit_changed();
        }
    }

    #[func]
    pub fn remove_path(&mut self, path: String) {
        let mut count = 0;

        for tag in self.tags.clone().as_slice() {
            if tag.to_string().starts_with(&path) {
                self.remove_tag(tag.to_owned());
                count += 1;
            }
        }

        if count > 0 {
            self.base_mut().emit_changed();
        }
    }

    #[func]
    pub fn some(&self, predicate: Callable) -> bool {
        self.tags.as_slice().iter().any(|tag| {
            let mut args = VariantArray::new();

            args.push(tag.to_variant());

            predicate.callv(args).booleanize()
        })
    }

    #[func]
    pub fn update_path(&mut self, old_path: String, new_path: String) {
        let mut count = 0;

        for tag in self.tags.clone().as_slice() {
            if tag.to_string().starts_with(&old_path) {
                let new_tag = tag.to_string().replace(&old_path, &new_path);
                self.replace_tag(tag.into(), new_tag.into());
                count += 1;
            }
        }

        if count > 0 {
            self.base_mut().emit_changed();
        }
    }
}

File: src/TagDictionary.cpp
//
// Created by iamoc on 17/03/2025.
//

#include "TagDictionary.h"

using namespace octod::gameplay::tags;

const char * TagDictionary::SPLIT_CHAR = ".";

bool TagDictionary::add_tag(const String &p_tag)
{
	if (has_tag(p_tag))
	{
		return false;
	}

	tags.append(p_tag);
	emit_changed();

	return true;
}

unsigned int TagDictionary::add_tags(const PackedStringArray &p_tags)
{
	unsigned int count = 0;

	for (int i = 0; i < p_tags.size(); i++)
	{
		if (add_tag(p_tags[i]))
		{
			count++;
		}
	}

	if (count > 0)
	{
		emit_changed();
	}

	return count;
}

unsigned int TagDictionary::count() const
{
	return tags.size();
}

PackedStringArray TagDictionary::find(const Callable &p_predicate) const
{
	PackedStringArray result;

	for (int i = 0; i < tags.size(); i++)
	{
		if (p_predicate.call(tags[i]))
		{
			result.append(tags[i]);
		}
	}

	return result;
}

PackedStringArray TagDictionary::get_tags_from_path(const String &p_path) const
{
	PackedStringArray result;

	for (int i = 0; i < tags.size(); i++)
	{
		if (tags[i].begins_with(p_path))
		{
			result.append(tags[i]);
		}
	}

	return result;
}

Dictionary TagDictionary::get_tree() const
{
	Dictionary root;
	PackedStringArray sorted_tags = tags;
	sorted_tags.sort();

	for (int i = 0; i < sorted_tags.size(); i++)
	{
		const String &tag = sorted_tags[i];
		PackedStringArray chunks = tag.split(".");
		Dictionary current = root;

		for (int j = 0; j < chunks.size(); j++)
		{
			const String &chunk = chunks[j];

			if (!current.has(chunk))
			{
				const Dictionary new_node;
				current[chunk] = new_node;
			}

			current = current[chunk];
		}
	}

	return root;
}

bool TagDictionary::has_path(const String &p_path) const
{
	for (int i = 0; i < tags.size(); i++)
	{
		if (tags[i].begins_with(p_path))
		{
			return true;
		}
	}

	return false;
}

bool TagDictionary::has_tag(const String &p_tag) const
{
	return tags.find(p_tag) != -1;
}

bool TagDictionary::has_some_tags(const PackedStringArray &p_tags) const
{
	for (int i = 0; i < p_tags.size(); i++)
	{
		if (tags.find(p_tags[i]) != -1)
		{
			return true;
		}
	}

	return false;
}

bool TagDictionary::has_none_tags(const PackedStringArray &p_tags) const
{
	for (int i = 0; i < p_tags.size(); i++)
	{
		if (tags.find(p_tags[i]) != -1)
		{
			return false;
		}
	}

	return true;
}

bool TagDictionary::none(const Callable &p_predicate) const
{
	for (int i = 0; i < tags.size(); i++)
	{
		if (p_predicate.call(tags[i]))
		{
			return false;
		}
	}

	return true;
}

bool TagDictionary::replace_tag(const String &p_old_tag, const String &p_new_tag)
{
	for (int i = 0; i < tags.size(); i++)
	{
		if (tags[i] == p_old_tag)
		{
			tags[i] = p_new_tag;
			emit_changed();
			return true;
		}
	}

	return false;
}

unsigned int TagDictionary::replace_tags(const PackedStringArray &p_old_tags, const PackedStringArray &p_new_tags)
{
	unsigned int count = 0;

	for (int i = 0; i < p_old_tags.size(); i++)
	{
		if (replace_tag(p_old_tags[i], p_new_tags[i]))
		{
			count++;
		}
	}

	if (count > 0)
	{
		emit_changed();
	}

	return count;
}

bool TagDictionary::remove_tag(const String &p_tag)
{
	const unsigned int tag_index = tags.find(p_tag);

	if (tag_index == -1)
	{
		return false;
	}

	tags.remove_at(tag_index);
	emit_changed();

	return true;
}

unsigned int TagDictionary::remove_path(const String &p_path)
{
	unsigned int removed_tags_count = 0;

	for (int64_t i = tags.size() - 1; i >= 0; i--)
	{
		if (tags[i].begins_with(p_path))
		{
			tags.remove_at(i);
			removed_tags_count++;
		}
	}

	if (removed_tags_count > 0)
	{
		emit_changed();
	}

	return removed_tags_count;
}

bool TagDictionary::some(const Callable &p_predicate) const
{
	for (int64_t i = 0; i < tags.size(); i++)
	{
		if (p_predicate.call(tags[i]))
		{
			return true;
		}
	}

	return false;
}

unsigned int TagDictionary::update_path(const String &p_old_path, const String &p_new_path)
{
	unsigned int updated_count = 0;

	for (int i = 0; i < tags.size(); i++)
	{
		if (tags[i].begins_with(p_old_path))
		{
			tags[i] = p_new_path + tags[i].substr(p_old_path.length());
			updated_count += 1;
		}
	}

	return updated_count;
}

void TagDictionary::_bind_methods()
{
	/// binds methods to godot
	ClassDB::bind_method(D_METHOD("add_tag", "string"), &TagDictionary::add_tag);
	ClassDB::bind_method(D_METHOD("find", "callable"), &TagDictionary::find);
	ClassDB::bind_method(D_METHOD("get_tags_from_path", "string"), &TagDictionary::get_tags_from_path);
	ClassDB::bind_method(D_METHOD("get_tree"), &TagDictionary::get_tree);
	ClassDB::bind_method(D_METHOD("has_path", "string"), &TagDictionary::has_path);
	ClassDB::bind_method(D_METHOD("has_tag", "string"), &TagDictionary::has_tag);
	ClassDB::bind_method(D_METHOD("has_some_tags", "array"), &TagDictionary::has_some_tags);
	ClassDB::bind_method(D_METHOD("has_none_tags", "array"), &TagDictionary::has_none_tags);
	ClassDB::bind_method(D_METHOD("none", "callable"), &TagDictionary::none);
	ClassDB::bind_method(D_METHOD("replace_tag", "string", "string"), &TagDictionary::replace_tag);
	ClassDB::bind_method(D_METHOD("replace_tags", "array", "array"), &TagDictionary::replace_tags);
	ClassDB::bind_method(D_METHOD("remove_tag", "string"), &TagDictionary::remove_tag);
	ClassDB::bind_method(D_METHOD("remove_path", "string"), &TagDictionary::remove_path);
	ClassDB::bind_method(D_METHOD("some", "callable"), &TagDictionary::some);
	ClassDB::bind_method(D_METHOD("update_path", "string", "string"), &TagDictionary::update_path);

	/// binds properties to godot
	ADD_PROPERTY(PropertyInfo(Variant::ARRAY, "tags", PROPERTY_HINT_TYPE_STRING), "add_tag", "remove_tag");
}
